[webdriver] Add initial tests: Get Active Element

jugglinmike opened the pull request:
Hi @andreastt! Just like with the patch for "Get Title", Firefox is failing all tests concerning unhandled prompt behavior. This means I haven't been able to validate those three except through visual inspection. I wanted to call this out so that you know to give those extra attention. Firefox passes all of the other tests.

<!-- Reviewable:start -->

<!-- Reviewable:end -->

jugglinmike committed:
[webdriver] Add initial tests: Get Active Element

w3c-bots commented:
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/237661794)

# Firefox (nightly) #
Testing web-platform-tests at revision 329782eec197717dea49557811d7adf37fa1a37e
Using browser at version BuildID 20170530100155; SourceStamp 286f71223256cbb3a769432fd860f563c4886e81
Starting 10 test iterations
All results were stable
## All results ##
<details>
<summary>1 test ran</summary>
<details>
<summary><a href="http://w3c-test.org/submissions/6098/webdriver/elements/active.py">/webdriver/elements/active.py</a></summary>


|                 Subtest                 | Results | Messages |
|-----------------------------------------|---------|----------|
|                                         | OK      |          |
| `active.py::test_closed_context`        | PASS    |          |
| `active.py::test_handle_prompt_dismiss` | FAIL    |          |

</details>
</details>

w3c-bots commented:
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/237661796)

# Sauce (safari) #

Using browser at version 10.0
Starting 10 test iterations
No tests run.

w3c-bots commented:
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/237661805)

# Chrome (unstable) #

w3c-bots commented:
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/237661809)

# Sauce (MicrosoftEdge) #

andreastt commented on webdriver/elements/active.py:
```diff
@@ -0,0 +1,210 @@
+            <input id="interactable-input" />
+            <input id="non-interactable-input" style="opacity: 0;" />
+            <p>Another element</p>
+        </body>""")
```
All the other items on the document are probably not relevant here, right? I get that you want to test that the first `<input>` is not selected by default but that feels like it’s testing HTML rather than WebDriver. HTML already defines that the default active element should be the document element, i.e. `<body>` in this case.

You probably want to test things like active element after interaction, frames &c. It’s of course fine to leave this until later.

andreastt reviewed (approved):
Thanks, this all looks good and can land as it is!

The tests as they are are fine, but you include a lot of HTML that is unnecessary to passing the tests. E.g. you don’t need `<body id="document-body">` or `<h1>`. In these cases you could just check that the active element has a tag name of `body` or use JS to compare the returned element to `document.activeElement`.

jugglinmike commented:
> I get that you want to test that the first `<input>` is not selected by
> default but that feels like it’s testing HTML rather than WebDriver.

On the one hand, I agree that the WebDriver tests for "Get Active Element" are
distinct from the HTML tests for `document.activeElement`. I'm tending to think
they are a superset, though, given that the former is defined in terms of the
latter. If an implementation of "Get Active Element" behaved differently than
`document.activeElement` for some edge case of the latter, than the
implementation would be incorrect.

What I think this means for these tests is that the edge cases should stay, but
all the expectations should be defined in terms of:

    execute_script('return document.activeElement;')

This is certainly weaker in some senses--if `document.activeElement` is wrong,
then the WebDriver tests won't catch it. But I suppose that's appropriate, given
what I've stated above. Passing WebDriver tests would be interpreted as, "the
WebDriver implementation faithfully reflects the current HTML behavior",
whereas passing HTML tests would mean, "the HTML implementation is correct."

> You probably want to test things like active element after interaction,
> frames &c. It’s of course fine to leave this until later.

I'm game to try this now.

andreastt commented:
FWIW I am absolutely fine with this PR landing as-is. Everything else discussed here can be iterated upon.

jugglinmike commented:
Thank you for that! Maybe I'm being a perfectionist, but since your suggestions were straightforward and I have the time now, I gave it a shot. Does this still look good to you?

andreastt commented on webdriver/elements/active.py:
```diff
@@ -0,0 +1,257 @@
+
+    assert_result_is_active_element(session, result)
+
+def test_sucess_without_body(session):
```
s/sucess/success/

andreastt commented on webdriver/elements/active.py:
```diff
@@ -0,0 +1,257 @@
+    assert_result_is_active_element(session, result)
+
+def test_success_iframe_content(session):
+    session.url = inline("<body></body>")
```
FWIW, `<body></body>`, and the use of the `<body>` tag in general, is unnecessary. (This is not a blocker.)

andreastt commented:
Still looks fine, modulo some nits.

jugglinmike commented:
Thanks for the review, @andreastt! I've rebased, squashed the `fixup!` commits, and force-pushed the result to this branch ([the previous version of the branch is available here](https://github.com/w3c/web-platform-tests/compare/master...bocoup:webdriver-get-active-element-prebase-2)). Would you mind merging on my behalf?
